src/run.rs

```diff
@@ -441,4 +447,35 @@ mod tests {
         let (_, fut) = run(cfg);
         fut.await.unwrap(); // Expect success as each op was retried
     }
+
+    struct AlwaysFailsOp(pub Option<Arc<Semaphore>>);
+
+    #[async_trait]
+    impl Operation for AlwaysFailsOp {
+        async fn execute(&mut self, _ctx: &OperationContext) -> Result<ControlFlow<()>> {
+            if let Some(s) = self.0.take() {
+                s.add_permits(1);
+            }
+            tokio::time::sleep(Duration::from_millis(10)).await; // Make sure we don't enter a spin loop
+            Err(anyhow::anyhow!("fail"))
+        }
+    }
+
+    #[tokio::test]
+    async fn test_ask_to_stop_on_constant_failures() {
+        let sem = Arc::new(Semaphore::new(0));
+        let sem_clone = Arc::clone(&sem);
+
+        let cfg = make_test_cfg(move || AlwaysFailsOp(Some(sem_clone.clone())));
+        let concurrency = cfg.concurrency as u32;
+
+        let (ctrl, fut) = run(cfg);
+
+        // Wait until all ops got stuck in retry loop
+        let _ = sem.acquire_many(concurrency).await.unwrap();
+
+        // Ask to stop and make sure that the workload finishes
+        ctrl.ask_to_stop();
+        fut.await.unwrap_err();
```

Comment:
We could add some sort of timeout so that the test fails instead of hanging.
Maybe something like:
```rust
tokio::time::timeout(tokio::time::Duration::from_secs(10) fut).await.expect("Test timed out").unwrap_err();
```
Or using the [`ntest`](https://github.com/becheran/ntest/tree/master/ntest_timeout) crate?

Reply:
Good point, I'll use the `ntest` crate.